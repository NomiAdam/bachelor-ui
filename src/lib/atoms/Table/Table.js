import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const StyledTable = styled.table`
	width: 100%;
	border-collapse: ${({rowSpacing}) => rowSpacing ? 'separate' : 'collapse'};
	padding-right: ${({rowSpacing}) => rowSpacing ? '5px' : '0'};
	border-spacing: ${({rowSpacing}) => rowSpacing ? '0 5px' : '0'};
	font-size: .85em;
`;

class Table extends PureComponent {
    render() {
        const {children, rowSpacing} = this.props;
        return (
            <StyledTable rowSpacing={rowSpacing}>
                {children}
            </StyledTable>
        )
    }
}

Table.propTypes = {
    /**
     * Children nodes of Table, Table body is expected
     */
    children: PropTypes.any,
    /**
     * Whether we should have space between rows or not
     */
    rowSpacing: PropTypes.bool,
};

Table.defaultProps = {
    rowSpacing: false,
};

export default Table;
