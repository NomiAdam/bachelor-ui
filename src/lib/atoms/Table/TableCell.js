import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledTableCell = styled.td`
	max-width: 100%;
	padding: 10px;
	${({editable}) => editable ? 'max-width: 100px' : ''};
	&:hover {
	  cursor: ${({redirect}) => redirect ? 'pointer' : 'inherit' };
	}
`;

class TableCell extends PureComponent {
	render() {
		const { children, colSpan, redirect, onClick, editable } = this.props;
		return(
			<StyledTableCell
				colSpan={colSpan}
				redirect={redirect}
                onClick={onClick}
                editable={editable}
			>
				{ children }
			</StyledTableCell>
		)
	}
}

TableCell.propTypes = {
	/**
	 * Number representation space between Cells
	 */
	colSpan:PropTypes.number,
	/**
	 * Children nodes of TableCell component
	 */
	children: PropTypes.any,
    /**
     * Whether we redirect on click or not
     */
    redirect: PropTypes.bool,
    /**
	 * Function to be called upon TableCell Click
     */
    onClick: PropTypes.func,
    /**
	 * Whether TableCell is used for Edit Icons
     */
    editable: PropTypes.bool,
};

TableCell.defaultProps = {
    redirect: false,
};

export { StyledTableCell };
export default TableCell;
