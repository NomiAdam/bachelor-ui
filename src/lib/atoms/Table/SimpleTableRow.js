import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { map, addIndex } from 'ramda';
import TableCell from './TableCell';

const mapIndexed = addIndex(map);

const StyledTableRow = styled.tr`
	background-color: transparent;
	td:first-child {
	  float: left;
	}
`;

const mapTableCells = (value, index) => <TableCell key={ index }>{value}</TableCell>;
const mappedTableCells = mapIndexed(mapTableCells);
class SimpleTableRow extends PureComponent {
    state = {
        rowData: [],
    };

    componentDidMount() {
        this.handleTableCellsMap();
    }

    componentDidUpdate(prevProps) {
        const { simpleTableRowData } = this.props;
        if (JSON.stringify(simpleTableRowData) !== JSON.stringify(prevProps.simpleTableRowData)) {
            this.handleTableCellsMap();
        }
    }

    handleTableCellsMap() {
        const { simpleTableRowData } = this.props;
        this.setState({ rowData: mappedTableCells(simpleTableRowData) });
    }

    render() {
        const { labeled, label } = this.props;
        const { rowData } = this.state;
        return (
            <StyledTableRow>
                {labeled && <TableCell>{label}</TableCell>}
                {rowData}
            </StyledTableRow>
        );
    }
}

SimpleTableRow.propTypes = {
    /**
     * Array of values is expected, value can be string, number or another Component
     */
    simpleTableRowData: PropTypes.array,
    /**
     * Boolean value, when true extra TabelCell is added to the row
     */
    labeled: PropTypes.bool,
    /**
     * Can be any value, be it string, number or Component, is displayed only when Row is labeled
     */
    label: PropTypes.any,
};

SimpleTableRow.defaultProps = {
    simpleTableRowData: [],
    labeled: false,
};

export { StyledTableRow };
export default SimpleTableRow;
