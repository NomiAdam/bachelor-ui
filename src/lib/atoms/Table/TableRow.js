import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { compose, tail, values } from 'ramda';
import { mapIndexed } from 'ramda-extension';
import { IoEdit, IoAndroidDelete } from 'react-icons/lib/io';
import TableCell from './TableCell';
import { lightTheme, darkTheme } from '../../constants/theme';

const StyledTableRow = styled.tr`
	background-color: transparent;
	border-bottom: 1px solid ${ lightTheme.lightGrey };
	transition: .25s all ease;
	&:hover {
		background-color: ${ ({ effect }) => (effect ? 'transparent' : lightTheme.whiteBlue) };
	}
`;

const Icon = styled.div`
    padding: 10px 15px;
    font-size: 1.5em;
    cursor: pointer; 
    color: ${ darkTheme.darkGrey };
`;

const StyledEdit = Icon.withComponent(IoEdit);
const StyledTrash = Icon.withComponent(IoAndroidDelete);

class TableRow extends PureComponent {
    state = {
        rowData: [],
    };

    componentDidMount() {
        this.handleTableCellsMap();
    }

    componentDidUpdate(prevProps) {
        const { tableRowData } = this.props;
        if (tableRowData !== prevProps.tableRowData) {
            this.handleTableCellsMap();
        }
    }

    handleTableCellsMap = () => {
        const { tableRowData, redirect, handleRedirect } = this.props;

        const mapTableCells = (value, index) => (
            <TableCell
                redirect={ redirect }
                key={ `${ index }-${ value }` }
                onClick={ handleRedirect }
            >
                {value}
            </TableCell>
        );
        const mappedTableCells = compose(tail, values, mapIndexed(mapTableCells));

        this.setState({ rowData: mappedTableCells(tableRowData) });
    };

    render() {
        const { editable, handleDelete, handleEdit } = this.props;
        const { rowData } = this.state;
        return (
            <StyledTableRow>
                {editable && (
                    <TableCell editable>
                        <StyledTrash onClick={ handleDelete } />
                        <StyledEdit onClick={ handleEdit } />
                    </TableCell>
                )}
                {rowData}
            </StyledTableRow>
        );
    }
}

TableRow.propTypes = {
    /**
     * Whether Table row should have something of Effect on hover, in our case color change
     */
    effect: PropTypes.bool,
    /**
     * Object of keys is expected
     */
    tableRowData: PropTypes.object,
    /**
     *  Whether table has extra column with edit Icons, boolean is expected
     */
    editable: PropTypes.bool,
    /**
     * Function to handle edit icon action, is present only when table is editable
     */
    handleEdit: PropTypes.func,
    /**
     * Function to handle delete icon action, is present only when table is editable
     */
    handleDelete: PropTypes.func,
    /**
     * Whether we redirect on click or not
     */
    redirect: PropTypes.bool,
    /**
     * Function to handle redirect on click, if redirect is set to false, handleRedirect is null
     */
    handleRedirect: PropTypes.any,
};

TableRow.defaultProps = {
    tableRowData: {},
    redirect: false,
    editable: false,
};

export { StyledTableRow, StyledEdit, StyledTrash };
export default TableRow;
