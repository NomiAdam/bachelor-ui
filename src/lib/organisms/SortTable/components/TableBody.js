import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {TableRow} from '../../../atoms/Table/index';
import {map, slice, o, dec, multiply, sort, prop, equals, ifElse, descend, ascend, always} from 'ramda';

const getSliceStart = o(multiply, dec);

const isAscending = equals('asc');
const sortByAscend = always(ascend);
const sortByDescend = always(descend);
const sortByOrder = ifElse(isAscending, sortByAscend, sortByDescend);

const getRowIdentification = prop('id');

class TableBody extends PureComponent {
    state = {
        columnData: [],
    };

    componentDidUpdate(prevProps) {
        const {
            columnData, order, page, rowsPerPage,
        } = this.props;
        if (
            columnData !== prevProps.columnData
            ||
            order !== prevProps.order
            ||
            page !== prevProps.page
            ||
            rowsPerPage !== prevProps.rowsPerPage
        ) {
            this.handleColumnDataMap();
        }
    };

    handleColumnDataMap = () => {
        const {
            columnData,
            rowsPerPage,
            page,
            orderBy,
            order,
            editable,
            handleDelete,
            handleRedirect,
            redirect,
            handleEdit
        } = this.props;

        const mapTableRow = (value) => (
            <TableRow
                tableRowData={value}
                key={getRowIdentification(value)}
                editable={editable}
                handleDelete={handleDelete(getRowIdentification(value))}
                handleEdit={handleEdit(getRowIdentification(value))}
                handleRedirect={redirect ? handleRedirect(getRowIdentification(value)) : null}
                redirect={redirect}
            />
        );

        const mappedTableRows = map(mapTableRow);

        const sliceStart = getSliceStart(page)(rowsPerPage);
        const sliceEnd = multiply(page, rowsPerPage);
        const sliceBody = slice(sliceStart, sliceEnd);

        const byOrder = sortByOrder(order)(prop(orderBy));

        const sortByGivenKey = sort(byOrder);
        const sortedBody = o(sortByGivenKey, sliceBody);

        const mappedBody = o(mappedTableRows, sortedBody);

        this.setState({ columnData: [] }, () => {
            this.setState({columnData: mappedBody(columnData)})
        });
    };

    render() {
        const {columnData} = this.state;
        return (
            <tbody>
            {columnData}
            </tbody>
        );
    };
}

TableBody.propTypes = {
    /**
     * Number representing current page
     */
    page: PropTypes.number,
    /**
     * Number representing how many rows per page are displayed
     */
    rowsPerPage: PropTypes.number,
    /**
     * String value, representing key, by which we will sort
     */
    orderBy: PropTypes.string,
    /**
     * String representation, either ascending, or descending
     */
    order: PropTypes.string,
    /**
     * [{id: 'id', key: 'value', anotherKey: 'val'}], note that every object must start with id key
     */
    columnData: PropTypes.array,
    /**
     * Function to handle current row delete
     */
    handleDelete: PropTypes.func,
    /**
     * Function to handle current row edit
     */
    handleEdit: PropTypes.func,
    /**
     * Boolean value, indicating whether the body is editable or not
     */
    editable: PropTypes.bool,
    /**
     * Whether we redirect on click or not
     */
    redirect: PropTypes.bool,
    /**
     * Function to handle redirect on click
     */
    handleRedirect: PropTypes.func,
};

TableBody.defaultProps = {
    columnData: [],
    editable: false,
    redirect: false,
};

export default TableBody;
