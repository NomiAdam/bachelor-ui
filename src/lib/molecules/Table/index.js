import TableCell from './TableCell';
import TableRow from './TableRow';
import SimpleTableRow from './SimpleTableRow';
import Table from './Table';

export {
    TableCell,
    TableRow,
    SimpleTableRow,
};

export default Table;
