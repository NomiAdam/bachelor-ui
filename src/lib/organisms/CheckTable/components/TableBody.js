import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import TableRow from './TableRow';
import {range, mapObjIndexed, values, o} from 'ramda';

class ReservationTableBody extends PureComponent {
    state = {
        bodyData: [],
    };

    componentDidUpdate(prevProps) {
        const { selected } = this.props;
        if( selected !== prevProps.selected ) {
            this.handleTableBodyMap();
        }
    }

    handleTableBodyMap = () => {
        const {rowRange, handleChange, selected, isCheckbox} = this.props;

        const cellRange = range(0, rowRange);

        const mapTableRow = ({value, label, disabled}, key) => (
            <TableRow
                key={key}
                label={label}
                dayHours={cellRange}
                selected={value}
                disabled={disabled}
                handleChange={handleChange(key)}
                isCheckbox={isCheckbox}
            />
        );

        const mapRow = mapObjIndexed(mapTableRow);
        const mappedTableRows = o(values, mapRow);

        this.setState({ bodyData: mappedTableRows(selected) })
    };

    render() {
        const { bodyData } = this.state;
        return (
            <tbody>
            {bodyData}
            </tbody>
        );
    }
}

ReservationTableBody.propTypes = {
    /**
     * Number indicating how many TableCells we will have in each row
     */
    rowRange: PropTypes.number,
    /**
     * Handle onClick on TableCell, (key) => (value) => is expected
     */
    handleChange: PropTypes.func,
    /**
     * Object of current table Rows, format as below
     * { key: { label: 'label', value: [] }, ... }
     * value array is array of selected (checked) TableCells
     */
    selected: PropTypes.object,
    /**
     * Whether we use checkbox component inside table or not
     */
    isCheckbox: PropTypes.bool,
};

ReservationTableBody.defaultProps = {
    dayHours: [],
    selected: {},
    rowRange: 0,
};

export default ReservationTableBody;
