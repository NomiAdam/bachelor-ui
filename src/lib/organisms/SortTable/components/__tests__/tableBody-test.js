/* eslint-disable no-undef */
import React from 'react';
import {shallow} from 'enzyme';
import TableBody from '../TableBody';
import {TableRow} from '../../../../atoms/Table/index';

const bodyData = [
    {
        key: 'Adam, Cell-1',
        another: 2,
        fuck: 'Row-1, Cell-3',
    },
    {
        key: 'Lukáš, Cell-1',
        another: 1,
    },
    {
        key: 'Anna, Cell-1',
        another: 6,
    },
    {
        key: 'Zikmund, Cell-1',
        another: 3,
    },
    {
        key: 'David, Cell-1',
        another: 8,
    },
    {
        key: 'Běta, Cell-1',
        another: 1,
    },
    {
        key: 'Karlos, Cell-1',
        another: 9,
    },
];

const mockPage = 1;
const mockRowsPerPage = 5;

const mockEdit = jest.fn();
const mockDelete = jest.fn();
const mockRedirect = jest.fn();

const renderTableBody = (
    <TableBody
        page={mockPage}
        rowsPerPage={mockRowsPerPage}
        order={'asc'}
        orderBy={'key'}
        handleEdit={mockEdit}
        handleDelete={mockDelete}
        handleRedirect={mockRedirect}
    />
);
const renderTableBodyEditable = (
    <TableBody
        page={mockPage}
        rowsPerPage={mockRowsPerPage}
        order={'asc'}
        orderBy={'key'}
        handleEdit={mockEdit}
        handleDelete={mockDelete}
        columnData={bodyData}
        handleRedirect={mockRedirect}
        editable
    />
);

describe('TableBody', () => {
    it('should be defined', () => {
        expect(TableBody).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(renderTableBody)).toMatchSnapshot();
    });
    it('should match the snapshot when editable', () => {
        expect(shallow(renderTableBodyEditable)).toMatchSnapshot();
    });
    it('should be of type tbody', () => {
        expect(shallow(renderTableBody).type()).toBe('tbody');
    });
    it('should have the same number of TableRows as are mockRowsPerPage', () => {
        const wrapper = shallow(renderTableBody);
        wrapper.setProps({ columnData: bodyData });
        expect(wrapper.find(TableRow).length).toBe(mockRowsPerPage);
    });
});
