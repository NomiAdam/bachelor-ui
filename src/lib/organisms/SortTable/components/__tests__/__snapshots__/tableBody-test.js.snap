// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableBody should match the snapshot 1`] = `<tbody />`;

exports[`TableBody should match the snapshot when editable 1`] = `<tbody />`;
